## Combinatorial Optimization


***A Tabu Search Algorithm with Short-term Memory for the Knapsack Problem - Solving Multiple Instances***

**Author:** Guilherme Cadori

**Date:** 24/08/2023


#### 1. Reading the instance

In [1]:
# Importing required libraries
import numpy as np

# Reading the target file
# GC: the code template below was provided in class and was adapted by GC (2023-07-09)

filename = "C:\gccadori\Lab5_CO_TabuSearch_STMOnly\Knap_C5I10.dat"

# Creating variables to store data
linha = None
coluna = None
cost = None
rhs = None
matrix = None

# Function to convert values in a line into a list of float values 
def parse_values(line):
    return [float(value) for value in line.strip().split()]

# Reading and extracting data
with open(filename, "r") as file:
    lines = file.readlines()

# Reading the scalar
linha = int(lines[0].strip())
coluna = int(lines[2].strip())

# Reading the vector
cost = np.array(parse_values(lines[5]))

# Reading the matrix
end = 7 + linha
matrix_lines = [line.strip() for line in lines[7:end]]
matrix = np.array([parse_values(line) for line in matrix_lines],dtype=object)

# Reading the RHS values
rhs = np.array(parse_values(lines[end+1]))

# Printing data for verification purposes
print("\n linhas:", linha)
print("\n colunas:", coluna)
print("\n cost:", cost)
print("\n rhs:", rhs)
print("\n Matrix:")
for row in matrix:
    print(row)



 linhas: 5

 colunas: 10

 cost: [0.30843 0.74907 0.6691  0.52619 0.61796 0.07105 0.16246 0.39898 0.57834
 0.19578]

 rhs: [116.12312  82.05387 109.88891  69.12963 128.29675]

 Matrix:
[13.2553 6.76342 22.86785 14.97223 13.27365 15.19261 17.27234 21.22448
 19.85637 0.93261]
[1.32876 7.51181 1.00781 13.34117 0.11301 24.33323 18.55862 14.78065
 18.35194 16.00987]
[17.40859 11.14715 0.07718 22.18725 1.06069 1.94107 23.637 17.02113
 24.08336 19.02182]
[0.27528 1.69006 4.87256 18.16527 3.74127 4.52415 12.46014 17.57698
 16.24719 16.45475]
[15.11814 15.51384 16.12041 10.67837 21.38512 19.73834 17.26438 12.49316
 22.49747 15.0008]


#### 2.	Implementing a Tabu Search with Short-Term Memory

In [2]:
def tabuSearchSTM(c, b, A, nonImprovingSolMax = 1, tabuSize = 1):
    
    # Importing auxiliary package
    import numpy as np
    
    # Function for creating an initial feasible solution
    def initialSolution(x, c, b, A):
        
        OF_Value = np.sum(x * c)
        
        feasibility = np.all(np.sum(x * A, 1) <= b)
        
        return OF_Value, feasibility
    
    # Function for generating neighbborhood
    def neighborGenerator(x):
       
        n = len(x)
        neighbors = []
        
        for i in range(n):
            neighbor = x.copy()
            neighbor[i] = 1 - neighbor[i]
            neighbors.append(neighbor)
       
        return neighbors
    
    # Creating search algorithm parameters
    x_best = []
    x_current = []
    tabuList = []
    feasibility = False
    n = len(c)
    randomNumber = np.random.default_rng()
    nonImproving = 0
    
    # Creating search algorithm main loop
    while feasibility == False:
        
        x_current = []
        
        for i in range(n):
            if randomNumber.random() < 0.5:
                x_current.append(1)
            else:
                x_current.append(0)

        x_best = x_current
        
        OF_bestValue, feasibility = initialSolution(x_current, c, b, A)

    
    while nonImproving <= nonImprovingSolMax:
        neighbors = neighborGenerator(x_current)

        x_bestNeighbor = None
        FO_Value_bestNeighbor = 0

        for neighbor in neighbors:
            if neighbor not in tabuList:
                neighborValue, neighborFeasbility = initialSolution(neighbor, c, b, A)
                if neighborValue > FO_Value_bestNeighbor and neighborFeasbility == True:
                    x_bestNeighbor = neighbor
                    FO_Value_bestNeighbor = neighborValue

        if x_bestNeighbor is None:
            break

        x_current = x_bestNeighbor
        tabuList.append(x_bestNeighbor)

        if len(tabuList) > tabuSize:
            tabuList.pop(0)

        if FO_Value_bestNeighbor > OF_bestValue:
            bestSolution = x_bestNeighbor
            OF_bestValue = FO_Value_bestNeighbor

        else:
            nonImproving = nonImproving + 1

    # Printing the best solution found
    print('Best solution found')
    print('Objective function value:', round(OF_bestValue, 2))
    
    return


In [3]:
# Testing the function
tabuSearchSTM(c = cost, 
              b = rhs, 
              A = matrix, 
              nonImprovingSolMax = coluna, 
              tabuSize = coluna)


Best solution found
Objective function value: 3.85


#### 3.	Running the heuristic for the given instance and returning the results

In [4]:
# Importing auxiliary package
import time

# Record starting time
start_time = time.time()

# Running the proposed heuristic for this instance
tabuSearchSTM(c = cost, 
              b = rhs, 
              A = matrix, 
              nonImprovingSolMax = coluna, 
              tabuSize = coluna)

# Record ending time
end_time = time.time()

# Calculate the elapsed time
elapsed_time = round(end_time - start_time, 4)

# Print the elapsed time
print("\nElapsed time:", elapsed_time, "seconds")


Best solution found
Objective function value: 3.85

Elapsed time: 0.0 seconds


**Test Run Results**

The best solution found:

- Objective function value: 3.85
- Elapsed time: 0.006 seconds

### End